### PR TITLE
Improve clang tidy to use compilation DB

### PR DESCRIPTION
**Pull request by veblush**
Inspired by https://github.com/envoyproxy/envoy/issues/921 and copied the script to build a compilation db from Envoy.

This PR improves the way to use clang-tidy which gRPC sanity test relies on to check if C++ sources have common errors. Because clang-tidy needs to be able to compile given sources, it has to have all necessary compilation options such as include directories. Previously those are provided manually by specifying all directories needed but this has limitations; i) someone needs to update this, ii) it's not always possible to specify include directory. e.g. open-census which doesn't have source files in third_party directory, and iii) clang-tidy prior to 7 silently ignores when it runs into the compilation errors which mostly are caused by insufficient include directories. As a result, clang-tidy hasn't been always run for all source files, which led degraded source code hygiene. 

A compilation DB can help to fix this problem. This DB includes all compilation options baked from Bazel so that clang-tidy can have enough options to compile given sources. This also helps open-census case because Bazel automatically fetches all necessary files to build targets so we don't need to have all source-files for clang-tidy. Simply saying, for all targets which Bazel can build, clang-tidy also can diagnose them. Finally, this can enable us to use clang-tidy 7 or later because all source files can be diagnosed successfully.

Note that I made some code changes by clang-tidy for files which clang-tidy wasn't able to run against.

### CODE REVIEW

**Comment by markdroth** [comment]
> Inspired by #921 and copied the script to build a compilation db from Envoy.

I don't think that's the right PR number.  Can you update it to point to the right thing?

File: tools/distrib/clang_tidy_code.sh [hunk 1]
```diff
@@ -21,6 +21,16 @@ set -ex
 cd $(dirname $0)/../..
 REPO_ROOT=$(pwd)
 
+# generate a clang compilation database
+tools/distrib/gen_compilation_database.py \
+  --include_headers \
+  --ignore_system_headers \
+  "//:*" \
+  "//src/core/..." \
```
Comment (markdroth) [inline tools/distrib/clang_tidy_code.sh:29]: Should this also cover `//src/cpp/...`?

File: tools/distrib/gen_compilation_database.py [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+    options = options.replace("-std=c++0x ", "")
+    options = options.replace("-std=c++11 ", "")
+
+    if args.vscode:
+        # Visual Studio Code doesn't seem to like "-iquote". Replace it with
+        # old-style "-I".
+        options = options.replace("-iquote ", "-I ")
+
+    if args.ignore_system_headers:
+        if target["file"] == "test/cpp/end2end/xds_end2end_test.cc":
```
Comment (markdroth) [inline tools/distrib/gen_compilation_database.py:94]: Why only for this one file?

File: tools/distrib/gen_compilation_database.py [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+            # Remove all include options for /usr/* directories
+            options = RE_INCLUDE_SYSTEM.sub("", options)
+
+    if isHeader(target["file"]):
+        options += " -Wno-pragma-once-outside-header -Wno-unused-const-variable"
+        options += " -Wno-unused-function"
+        if not target["file"].startswith("external/"):
+            # *.h file is treated as C header by default while our headers files are all C++17.
```
Comment (markdroth) [inline tools/distrib/gen_compilation_database.py:102]: Why is this needed?  And why C++17?  Aren't we still requiring C++11 support?

File: tools/run_tests/sanity/sanity_tests.yaml [hunk 4]
```diff
@@ -24,6 +24,7 @@
 - script: tools/distrib/check_pytype.sh
 - script: tools/distrib/clang_format_code.sh
 - script: tools/distrib/clang_tidy_code.sh
+  cpu_cost: 1000
```
Comment (markdroth) [inline tools/run_tests/sanity/sanity_tests.yaml:27]: What does this do?

**Review by markdroth — COMMENTED** [review]
Thanks for doing this!

File: tools/distrib/clang_tidy_code.sh (hunk 1 above)
Comment (veblush) [reply tools/distrib/clang_tidy_code.sh:29]: I like to do it although there are no targets under //src/cpp/... but blaze throws an error if the expression doesn't have any.

File: tools/distrib/gen_compilation_database.py (hunk 2 above)
Comment (veblush) [reply tools/distrib/gen_compilation_database.py:94]: Oops my test code got in accidentally.

File: tools/run_tests/sanity/sanity_tests.yaml (hunk 4 above)
Comment (veblush) [reply tools/run_tests/sanity/sanity_tests.yaml:27]: This is to make clang-tidy run exclusively when running on Kokoro machine. This is required to run bazel without being interrupted by other tasks.

File: tools/distrib/gen_compilation_database.py (hunk 3 above)
Comment (veblush) [reply tools/distrib/gen_compilation_database.py]: Yeah C++11 would be the right one

**Comment by veblush** [comment]
Mark, thanks for the quick review! I updated the description and code accordingly.

File: tools/run_tests/sanity/sanity_tests.yaml (hunk 4 above)
Comment (markdroth) [reply tools/run_tests/sanity/sanity_tests.yaml:27]: Maybe add a comment here explaining this?

**Review by markdroth — APPROVED** [review]
This looks good to me, but please get either Nico or Jan to approve before merging.  Thanks!

**Comment by veblush** [comment]
Thanks, Mark! @nicolasnoble If you have a chance, can you review this?

File: tools/run_tests/sanity/sanity_tests.yaml (hunk 4 above)
Comment (veblush) [reply tools/run_tests/sanity/sanity_tests.yaml:27]: Good idea. This is a yaml file so I should leave comments.

File: tools/distrib/clang_tidy_code.sh [hunk 5]
```diff
@@ -21,6 +21,22 @@ set -ex
...
+# grep targets with manual tag, which is not included in a result of bazel build using ...
+# let's get a list of them using query command and pass it to gen_compilation_database.py
+export MANUAL_TARGETS=$(bazel query 'attr("tags", "manual", tests(//test/cpp/...))' | grep -v _on_ios)
+
+# generate a clang compilation database
+tools/distrib/gen_compilation_database.py \
+  --include_headers \
+  --ignore_system_headers \
+  --dedup_targets \
+  "//:*" \
```
Comment (jtattermusch) [inline tools/distrib/clang_tidy_code.sh:33]: I wish there was a better way of identifying the targets we want to include in compdb (I'd like something that's zero-maintenance), but I'm assuming you already explored that option and could not come up with anything better (but still simple).
Perhaps add a TODO that mentions what would ideally be the set of targets we'd like to list here (is it "all C/C++ code in this repo"?), and that the current list of targets is the best approximation of that we could get.

File: tools/run_tests/sanity/sanity_tests.yaml [hunk 6]
```diff
@@ -24,6 +24,9 @@
 - script: tools/distrib/check_pytype.sh
 - script: tools/distrib/clang_format_code.sh
 - script: tools/distrib/clang_tidy_code.sh
+  # ClangTidy needs to run exclusively because it runs bazel making changes
+  # and it needs many CPUs to parse all source files to be diagnosed.
+  cpu_cost: 1000
```
Comment (jtattermusch) [inline tools/run_tests/sanity/sanity_tests.yaml:29]: I don't understand why running clang tidy as part of sanity check would need to modify anything (under sanity test, it should only try to clang tidy files in a separate subtree and then diff whether everything looks the same as the checked-in code). This should not collide with anything (same behavior as before).

Therefore I don't understand why we would need to run exclusively - that seems wrong, please explain.

Comment (jtattermusch) [inline tools/run_tests/sanity/sanity_tests.yaml:29]: What is the average clang-tidy check duration? We want to make our sanity check run as fast as possible.

**Review by jtattermusch — CHANGES_REQUESTED** [review]

File: tools/run_tests/sanity/sanity_tests.yaml (hunk 6 above)
Comment (veblush) [reply tools/run_tests/sanity/sanity_tests.yaml:29]: This couldn't be faster by design. It should compile all C++ header and source files to draw a conclusion. It usually takes 7-8 minutes on my gLinux desktop.

File: tools/distrib/clang_tidy_code.sh (hunk 5 above)
Comment (veblush) [reply tools/distrib/clang_tidy_code.sh:33]: Yeah, it's not easy and probably not possible to get it fully automatic. I updated the comment slightly to explain what's done here.

File: tools/run_tests/sanity/sanity_tests.yaml (hunk 6 above)
Comment (veblush) [reply tools/run_tests/sanity/sanity_tests.yaml:29]: Let me retry this. It wasn't able to run deterministically before without this. I agree with you and it won't change anything except `compile_commands.json` but it didn't run successfully without this. Let me retry this.

Comment (veblush) [reply tools/run_tests/sanity/sanity_tests.yaml:29]: Oh I figured this out. This is still required because clangtidy needs files under the bazel output directory because the compilation database is built from bazel and it builds a new source directory view in that directory. (This is why it can access third_party files without having third_party sources in the repo) But this view can be removed or changed by another bazel invocation so it should run exclusively. I updated the comment.

Comment (veblush) [reply tools/run_tests/sanity/sanity_tests.yaml:29]: Running a couple of bazel at the same time would be fine. In this case, however, running bazel to build a compilation db and running clangtidy based on the db which have links to files in the bazel output directory. When another bazel build runs, it will change that directory while clangtidy is still working on it.

**Review by jtattermusch — APPROVED** [review]
LGTM.
